### ISSUE TITLE
Multiple Node Projects Single WORKSPACE FILE

### ISSUE THREAD

**Issue by jried31** [angular/angular-bazel-example#456]
We have a mono-repo, where there are many programming languages in it such as Java, Python, etc. I'd like to create an Angular project, however it seems as though the package.json file and the`yarn_install` command are tied to the WORKSPACE file. To my knowledge, means that there is a single package.json file for all Node projects in the repo. 

Node projects (incl Angular) naturally define their individual dependancies inside of a package.json file. Is there a way to do this with Bazel's managed dependancies, while preservin a single WORKSPACE file and allowing various Node type projects to exist independantly (eg: Node, Angular, etc)?

In other words, I'd like to see something like the following where each project can exist independently without having to place project specific details in the WORKSPACE file:

web.something.nodejs
--- package.json
--- BUILD

web.something.angular
--- package.json
--- BUILD

web.common
--- BUILD

**Comment by alexeagle** [comment]
rules_nodejs, angular-cli, and angular/angular are all examples of Bazel monorepo with one WORKSPACE file and many node_modules directories.

our simple example has a single `yarn_install` in the WORKSPACE file because we find it convenient to put all the devDependencies together in the top-level file. However you could totally have multiple `yarn_install` commands so that different subdirectories use different bazel workspaces for the node_modules.

Note that you won't be able to use the `@npm` repository name for more than one directory, so you'll have to be more explicit in your BUILD files about which node_modules directory should be used as input to each rule.

**Comment by jried31** [comment]
Could you point me to examples where this is successfully done? We have multiple node based projects (angular being one) and want to construct it s.t. there is a single WORKSPACE file and each project maintains its own project specific dependancies.

**Comment by alexeagle** [comment]
Sure, here's an example:
https://github.com/alexeagle/demo-bazel-multipackage

note that any given program or build step (action) can only load dependencies from a single node_modules at a time. This is different from eg. yarn workspaces where shared deps can be hoisted to a node_modules in a subdirectory.

**Comment by jried31** [comment]
@alexeagle, your the man! Thank you :)

**Comment by gregmagolan** [comment]
You'll run into some issues with rules such as ts_devserver as they have implicit attributes that assume `@npm`. For these would setup a `defaults.bzl` (something like  https://github.com/angular/angular/blob/master/tools/defaults.bzl) in each sub-project that overrides these attributes to your `@a_deps` / `@b_deps` etc and re-exports the rules for use in that sub-project... @alexeagle We should example this example to include that.

**Comment by jried31** [comment]
@alexeagle or @gregmagolan  --- what min version of Bazel does this solution work with? It's not defined in the WORKSPACE and we're using 0.9.0.

**Comment by gregmagolan** [comment]
Minimum version for rules_nodejs is 0.26.0. Although you can go back to 0.21.0 if you opt out of symlinked node_modules by setting `symlink_node_modules = False` in your yarn_install rules.

**Comment by alexeagle** [comment]
sounds like this was resolved (though I think we could do a better job explaining how Bazel solves the monorepo use case)
